File: q002_add_two_nums.py
#!/usr/bin/env python
# coding: utf-8

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        a=l1
        b=l2
        result = c = ListNode(None)
        carry=0
        while a or b:
            av = a.val if a else 0
            bv = b.val if b else 0
            x = av + bv + carry
            c.next = ListNode(x % 10)
            carry = 1 if x >= 10 else 0
            if a: a = a.next
            if b: b = b.next
            c = c.next
        
        if carry: c.next = ListNode(carry)
        
        return result.next


def list2listnode(nums):
    lstnode = ListNode(None)
    curr_node = lstnode
    for n in nums:
        curr_node.next = ListNode(n)
        curr_node = curr_node.next
    return lstnode.next


def listnode2list(lstnode):
    lst = []
    while lstnode:
        lst.append(lstnode.val)
        lstnode = lstnode.next
    return lst


if __name__ == "__main__":
    testcases=[
        ([2, 4, 3], [5,6,4], [7,0,8]),
        ([0], [0], [0]),
        ([1, 2, 7], [8,9], [9, 1, 8]),

    ]

    solution = Solution()
    for i, testcase in enumerate(testcases):
        ans = solution.addTwoNumbers(list2listnode(testcase[0]), list2listnode(testcase[1]))
        print(i, 'OK' if listnode2list(ans) == testcase[2] else 'Failed', 'expected:%s, return:%s'%(testcase[2], listnode2list(ans)))

# Runtime: 68 ms, faster than 93.39% of Python online submissions.

File: q002_add_two_nums.cpp
 #include <vector>
#include <iostream>
#include <string>
#include <tuple>

using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x=0, ListNode *next=nullptr) : val(x), next(next) {}
    ListNode(const vector<int>& v): ListNode()
    {
        auto p = this;
        for (const auto&i : v)
        {
            p->next = new ListNode(i);
            p = p->next;
        }
    }

    vector<int> toVector() const
    {
        vector<int> v;
        auto p = this;
        while (p) {
            v.emplace_back(p->val);
            p = p->next;
        }
        return v;
    }

};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode ret;
        auto a = l1;
        auto b = l2;
        auto c = &ret;
        int carry = 0;

        while (a or b)
        {
            int av = a ? a->val : 0;
            int bv = b ? b->val : 0;
            int x = av + bv + carry;
            c->next = new ListNode(x % 10);
            carry = x >= 10 ? 1 : 0;
            if (a) {
                a = a->next;
            }
            if (b) {
                b = b->next;
            }
            c = c->next;
        }

        if (carry) {
            c->next = new ListNode(carry);
        }

        return ret.next;
    }
};


auto print_all = [](const auto &v, const string &s="") {
    if (s != "") cout << s;
    for (const auto &i:v) cout << i << ' ';
    cout << endl;
};

int main(int argc, char const *argv[])
{
    vector<tuple<vector<int>, vector<int>, vector<int>>> testCases {
        {{2,4,3}, {5,6,4}, {7,0,8}},
        {{0    }, {0    }, {0    }},
        {{1,2,7}, {8,9  }, {9,1,8}},
        {{9,9,9,9,9,9,9}, {9,9,9,9},{8,9,9,9,0,0,0,1}}
    };
    auto s = Solution();
    for (auto& testcase:testCases){
        auto idx = &testcase - testCases.data();
        ListNode l1(std::get<0>(testcase));
        ListNode l2(std::get<1>(testcase));
        auto l3_p = s.addTwoNumbers(l1.next, l2.next);
        auto ret = l3_p->toVector();
        cout << "case " << idx << ": " <<(ret == std::get<2>(testcase) ? "PASSED" : "FAILED") << endl;
        print_all(std::get<2>(testcase), "    expected: ");
        print_all(ret, "    returned: ");
    }

    return 0;
}
